providers/providers.go

```diff
@@ -4,26 +4,31 @@ import (
 	"context"
 
 	"github.com/coreos/go-oidc"
+	mw "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/middleware"
 	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
 )
 
 // Provider represents an upstream identity provider implementation
 type Provider interface {
 	Data() *ProviderData
 	// DEPRECATED: Migrate to EnrichSessionState
 	GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error)
-	Redeem(ctx context.Context, redirectURI, code string) (*sessions.SessionState, error)
+	Redeem(ctx context.Context, ps mw.ProxyState, redirectURI, code string) (*sessions.SessionState, error)
```

Comment:
I don't particularly like `ProxyState` needing to be added to all these methods.

Do you think it is still needed based on my other comments?

If so, I think this belongs on `ProviderData` struct since it is static between distinct sessions/requests.

Reply:
This is not *provider* data. This is to allow providers to query exposed common proxy configuration. We can do it through `ProxyConfig` similar to `ProviderData` to allow providers to access the common proxy settings (I find the current inability to do so a peculiar choice). 

All these implementation decisions were made to follow the existing patterns and provide minimal intrusion and maximal backwards compatibility. I didn't want to "rewrite" and if I did I wouldn't have written a lot of things the same way.

Reply:
Do you even need it if you handle all the timing & refresh logic in the middleware `Refresh` which chooses whether or not to refresh based on the Proxy config options (which that layer already has access to)?

I really want to keep these `provider` interfaces clean and limit the arguments so they are easy to understand & extend for other contributors. `context.Context` & `sessions.SessionState` for the most part.

Reply:
If the some custom provider logic relies on global settings, you need to know what settings user set for the proxy from inside the provider. `ProxyConfig` can be exposed through `Provider` interface to as `Data()` and leave the methods Provider method signatures alone.

Reply:
What I'm getting at is I don't see a scenario where providers need this data if we implement the logic globally in the middleware that calls `RefreshSession`.

If they ever did, any special logic they have on timing they can pass on by setting the `CreatedAt` or `ExpiresOn` -- but I'm not seeing any use cases in this PR that would need the provider to know the refresh timing details (assuming it is now `RefreshSession` and not `RefreshSessionIfNeeded`.

Off the top of my head, I don't think any of the other providers that have implemented `RefreshSessionIfNeeded` need to do anything special either.